### PR TITLE
MAGECLOUD-4970: Move all environment variables to schema config

### PR DESCRIPTION
**Pull request by oshmyheliuk**
### Description

https://magento2.atlassian.net/browse/MAGECLOUD-4970

### Manual testing scenarios

1. Run `php bin/ece-tools schema:generate` comnmand
2. Check that all configuration options exist in the dumped file `.magento.env.md`

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: config/schema.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Stage variables
 VERBOSE_COMMANDS:
-  description: Adjust the output verbosity level of Magento commands.
+  description: "Adjust the output verbosity level of Magento commands."
```
Comment (shiftedreality) [inline config/schema.yaml]: There is no need in quotes - https://yaml.org/spec/1.2/spec.html#id2760844

Comment (shiftedreality) [reply config/schema.yaml]: http://blogs.perl.org/users/tinita/2018/03/strings-in-yaml---to-quote-or-not-to-quote.html

Comment (oshmyheliuk) [reply config/schema.yaml]: Some description messages contain colon symbol `:` and without quotes, `:<space>` won't work.
I just changed all description messages to one style.

Comment (shiftedreality) [reply config/schema.yaml]: This will be lost on the dump in any case, please remove colons where they’re not needed

Comment (oshmyheliuk) [reply config/schema.yaml]: Colon won't be lost on the dump. We have a lot of variables with such construction `_merge: true` in the description. Also, there is one message which requires newlines, without quotes it won't work too.

Comment (shiftedreality) [reply config/schema.yaml]: Quotes will be lost, not colons. Yaml Dump class by default follows yaml spec and won’t dump quotes

Comment (oshmyheliuk) [reply config/schema.yaml]: Yes, I understand that quotes will be lost. But as I mentioned some descriptions won't work without quotes, and I just changed all descriptions to one format.

File: config/schema.yaml [hunk 2]
```diff
@@ -281,6 +283,299 @@ CLEAN_STATIC_FILES:
...
+    - deploy
+  default:
+    deploy: []
+  examples:
+    - comment: Default value depends on template configuration of ElasticSearch (i.e. default number of replicas and shards)
+        and ElasticSearch host.
+      stage:
+        deploy:
+          ELASTICSUITE_CONFIGURATION:
+            _merge: false if true configuration merging with default"
```
Comment (billygilbert) [inline config/schema.yaml]: If this is an example, it should just be true/false.  Also looks like an extra "

File: config/schema.yaml [hunk 3]
```diff
@@ -281,6 +283,299 @@ CLEAN_STATIC_FILES:
...
+        deploy:
+          CONSUMERS_WAIT_FOR_MAX_MESSAGES: true
+ENABLE_GOOGLE_ANALYTICS:
+  description: Prevents the disabling of Google Analytics.  By default, Google Analytics is enabled only
+    on Production environments. Set this value to true to enable Google Analytics on other environments as well.
+    Google Analytics is always enabled in Production.
+  type: boolean
+  stages:
+    - deploy
+    - global
```
Comment (billygilbert) [inline config/schema.yaml:529]: Why do we note global here and not other places (example line 508)

File: config/schema.yaml [hunk 4]
```diff
@@ -281,6 +283,299 @@ CLEAN_STATIC_FILES:
...
+    - stage:
+        deploy:
+          GENERATED_CODE_SYMLINK: false
+WARM_UP_PAGES:
+  description: Customize the list of pages used to preload the cache in the post_deploy stage.
+  type: array
+  stages:
+    - post-deploy
+  default:
+    post-deploy: [""]
```
Comment (billygilbert) [inline config/schema.yaml:554]: should the "" be there?  Different than 560 which is a similar config.

File: config/schema.yaml [hunk 5]
```diff
@@ -281,6 +283,299 @@ CLEAN_STATIC_FILES:
...
+  examples:
+    - stage:
+        post-deploy:
+          WARM_UP_PAGES:
+            - "index.php"
+            - "index.php/customer/account/create"
```
Comment (billygilbert) [inline config/schema.yaml:560]: Should we have an example here showing full URL? IE https://example.com/checkout/cart

File: config/schema.yaml [hunk 6]
```diff
@@ -338,3 +633,21 @@ ENV_ENVIRONMENT:
     - system:
         variables:
           ENV_ENVIRONMENT: MAGENTO_CLOUD_ENVIRONMENT
+
+# Dev variables
```
Comment (billygilbert) [inline config/schema.yaml:645]: I believe these variables are not technically supported outside of our internal teams, I don't know if we want to document them.  @YPyltiai, thoughts?

**Review by billygilbert — CHANGES_REQUESTED** [review]

File: config/schema.yaml (hunk 2 above)
Comment (shiftedreality) [reply config/schema.yaml]: String, where allowed value is boolean, may be confusing, possibly better to move to comment section

File: config/schema.yaml [hunk 7]
```diff
@@ -281,6 +283,299 @@ CLEAN_STATIC_FILES:
...
+  type: array
+  stages:
+    - deploy
+  default:
+    deploy: []
+  examples:
+    - stage:
+        deploy:
+          CACHE_CONFIGURATION:
+            _merge: false if true configuration merging with default
```
Comment (shiftedreality) [inline config/schema.yaml]: Same here, these options should be boolean

File: src/Command/GenerateSchema.php [hunk 8]
```diff
@@ -82,9 +91,13 @@ public function execute(InputInterface $input, OutputInterface $output)
...
         $data = $this->schema->getSchema();
 
+        $this->logger->info('Starting schema dist file generation');
+
         $this->file->filePutContents(
             $this->fileList->getEnvDistConfig(),
             $this->formatter->format($data)
         );
+
+        $this->logger->info(sprintf('Dist file was successfully generated: %s', $this->fileList->getEnvDistConfig()));
```
Comment (shiftedreality) [inline src/Command/GenerateSchema.php]: Logging in command classes does not have much value, Command output should be used instead

File: src/Config/Schema/Formatter.php [hunk 9]
```diff
@@ -39,14 +39,18 @@ public function format(array $data): string
         $text = '';
 
         foreach ($data as $key => $item) {
+            if (isset($item[Schema::SCHEMA_DUMP]) && !$item[Schema::SCHEMA_DUMP]) {
```
Comment (shiftedreality) [inline src/Config/Schema/Formatter.php]: `isset()` should be eliminated in favor to `!empty()` only

**Review by shiftedreality — CHANGES_REQUESTED** [review]

File: config/schema.yaml (hunk 3 above)
Comment (oshmyheliuk) [reply config/schema.yaml:529]: Moved from old `Schema.php` file without changes to avoid BIC

File: config/schema.yaml (hunk 4 above)
Comment (oshmyheliuk) [reply config/schema.yaml:554]: Yes, this is a specific of this particular configuration. It moved from `Schema.php` too.

File: config/schema.yaml (hunk 6 above)
Comment (oshmyheliuk) [reply config/schema.yaml:645]: Dev variables won't dump to the documentation if they have an option: `dump: false`.
But we need them in schema.yaml for the ability to use them in the code.

File: src/Config/Schema/Formatter.php (hunk 9 above)
Comment (oshmyheliuk) [reply src/Config/Schema/Formatter.php]: `!empty` doesn't work in this case. We need to check if option configured and it `false`

File: config/schema.yaml (hunk 5 above)
Comment (oshmyheliuk) [reply config/schema.yaml:560]: Example added

File: config/schema.yaml (hunk 7 above)
Comment (oshmyheliuk) [reply config/schema.yaml]: Moved `if true configuration merging with default` to the example comment section

File: src/Config/Schema/Formatter.php (hunk 9 above)
Comment (oshmyheliuk) [reply src/Config/Schema/Formatter.php]: done

File: config/schema.yaml [hunk 10]
```diff
@@ -311,10 +311,11 @@ SEARCH_CONFIGURATION:
   default:
     deploy: []
   examples:
-    - stage:
+    - comment: If `_merge` is true configuration merging with default
```
Comment (YPyltiai) [inline config/schema.yaml]: `configuration merging with default` -> `configuration is being merged with the default`

File: config/schema.yaml (hunk 6 above)
Comment (YPyltiai) [reply config/schema.yaml:645]: If they are not going to Docs - Sasha's explanation should work.

File: src/Command/GenerateSchema.php (hunk 8 above)
Comment (oshmyheliuk) [reply src/Command/GenerateSchema.php]: Changed

File: config/schema.yaml (hunk 10 above)
Comment (oshmyheliuk) [reply config/schema.yaml]: Thanks, changed.

**Review by billygilbert — DISMISSED** [review]

**Review by shiftedreality — DISMISSED** [review]

**Review by billygilbert — APPROVED** [review]

**Review by shiftedreality — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request merged**